File: app/src/main/java/com/example/firabasefirstexperience/Activities/ChatActivity.java
package com.example.firabasefirstexperience.Activities;

import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import com.example.firabasefirstexperience.Adapter.ChatAdapter;
import com.example.firabasefirstexperience.Model.Chat;
import com.example.firabasefirstexperience.Model.User;
import com.example.firabasefirstexperience.R;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;

public class ChatActivity extends AppCompatActivity {

    private CircleImageView profileImage;
    private TextView tv_UserName;
    private ImageButton but_Send;
    private EditText et_Message;

    private FirebaseUser fb_User;
    private DatabaseReference reference;

    private Intent intent;

    private ChatAdapter chatAdapter;
    private List<Chat> listOfChats;

    private RecyclerView recView_Chats;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);

        profileImage = findViewById(R.id.profileImage);
        tv_UserName = findViewById(R.id.tv_UserName);
        but_Send = findViewById(R.id.but_Send);
        et_Message = findViewById(R.id.et_Message);


        Toolbar toolbar = findViewById(R.id.toolBar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle("");
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });

        recView_Chats = findViewById(R.id.recView_Chats);
        recView_Chats.setHasFixedSize(true);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getApplicationContext(), RecyclerView.VERTICAL, false);
        linearLayoutManager.setStackFromEnd(true);
        recView_Chats.setLayoutManager(linearLayoutManager);


        intent = getIntent();

        final String userId = intent.getStringExtra("userId");

        but_Send.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String msg = et_Message.getText().toString();
                if (!msg.isEmpty()) {
                    sendMessage(fb_User.getUid(), userId, msg);
                } else {
                    Toast.makeText(ChatActivity.this, "You can't send an empty message", Toast.LENGTH_SHORT).show();
                }

                et_Message.setText(null);

            }
        });


        fb_User = FirebaseAuth.getInstance().getCurrentUser();

        reference = FirebaseDatabase.getInstance().getReference("Users").child(userId);


        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                User user = dataSnapshot.getValue(User.class);

                tv_UserName.setText(user.getUserName());

                readMessage(fb_User.getUid(), userId);
            }


            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }


    private void sendMessage(String sender, String receiver, String message) {

        DatabaseReference reference = FirebaseDatabase.getInstance().getReference();

        HashMap<String, Object> hm_Messagee = new HashMap<>();
        hm_Messagee.put("sender", sender);
        hm_Messagee.put("receiver", receiver);
        hm_Messagee.put("message", message);


        reference.child("Chats").push().setValue(hm_Messagee);
    }

    private void readMessage(final String myId, final String userId) {
        listOfChats = new ArrayList<>();

        reference = FirebaseDatabase.getInstance().getReference("Chats");

        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                listOfChats.clear();

                for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                    Chat chat = snapshot.getValue(Chat.class);

                    if (chat.getReceiver() != null) {

                        if ((chat.getReceiver().equals(myId) && chat.getSender().equals(userId)) || (chat.getReceiver().equals(userId) && chat.getSender().equals(myId))) {
                            listOfChats.add(chat);

                            chatAdapter = new ChatAdapter(ChatActivity.this, listOfChats);
                            recView_Chats.setAdapter(chatAdapter);
                        }
                    }


                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }


}

File: app/src/main/java/com/nassdk/supchat/presentation/chat/ui/ChatActivity.kt
package com.nassdk.supchat.presentation.chat.ui

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.arellomobile.mvp.MvpAppCompatActivity
import com.arellomobile.mvp.presenter.InjectPresenter
import com.bumptech.glide.Glide
import com.nassdk.supchat.R
import com.nassdk.supchat.presentation.chat.MessageAdapter
import com.nassdk.supchat.model.Chat
import com.nassdk.supchat.model.User
import com.nassdk.supchat.presentation.chat.mvp.ChatActivityPresenter
import com.nassdk.supchat.presentation.chat.mvp.ChatActivityView
import com.nassdk.supchat.presentation.diffprofile.ui.DiffProfileActivity
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.*
import kotlinx.android.synthetic.main.activity_chat.*
import java.util.ArrayList

class ChatActivity : MvpAppCompatActivity(), ChatActivityView {

    private lateinit var fbUser: FirebaseUser
    private lateinit var reference: DatabaseReference

    private lateinit var messageAdapter: MessageAdapter
    private lateinit var listOfChat: List<Chat>

    private lateinit var recViewChats: RecyclerView

    @InjectPresenter
    lateinit var presenter: ChatActivityPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat)

        initViews()
    }


    private fun initViews() {
        setSupportActionBar(toolBar)
        supportActionBar?.title = ""
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        toolBar?.setNavigationOnClickListener { finish() }

        recViewChats = findViewById(R.id.recView_Chats)
        recViewChats.setHasFixedSize(true)
        val linearLayoutManager = LinearLayoutManager(applicationContext, RecyclerView.VERTICAL, false)
        linearLayoutManager.stackFromEnd = true
        recView_Chats?.layoutManager = linearLayoutManager

        val intent = intent

        val userId = intent.getStringExtra("userId")

        fbUser = FirebaseAuth.getInstance().currentUser!!
        reference = FirebaseDatabase.getInstance().getReference("Users").child(userId)

        but_Send.setOnClickListener {
            val msg = et_Message.text.toString()
            if (msg.isEmpty()) {
                presenter.onEmptyError()
            } else {
                if (presenter.checkInternetConnection(context = applicationContext)) {
                } else {
                    presenter.sendMessage(fbUser.uid, userId, msg)
                }
            }

            et_Message.text = null
        }

        tv_UserName.setOnClickListener {
            presenter.toDiffProfile()
        }

        reference.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {

                val user = dataSnapshot.getValue(User::class.java)

                if (user != null) {
                    tv_UserName.text = user.userName

                    if (user.imageURL == "default") {
                        profileImage_Chat.setImageResource(R.mipmap.ic_launcher_round)
                    } else {
                        Glide
                                .with(applicationContext)
                                .load(user.imageURL)
                                .into(profileImage_Chat)
                    }
                }

                if (user != null) {
                    presenter.readMessage(fbUser.uid, userId)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {
            }
        })
    }

    override fun showEmptyError() {
        Toast.makeText(applicationContext, "You can't send an empty Message", Toast.LENGTH_SHORT).show()
    }

    override fun toDiffProfile() {
        val intent = intent
        val userId = intent.getStringExtra("userId")

        val intentToDiffProfile = Intent(applicationContext, DiffProfileActivity::class.java)
        intentToDiffProfile.putExtra("id", userId)
        startActivity(intentToDiffProfile)
    }

    override fun setAdapter(listOfChats: ArrayList<Chat>) {
        messageAdapter = MessageAdapter(listOfChats)
        recView_Chats?.adapter = messageAdapter
    }

    override fun showDialog() {
        val builder = AlertDialog.Builder(this@ChatActivity)
        builder.setTitle("Warning!")
                .setMessage("Your device is not connected to Internet. Please, try later")
                .setIcon(R.drawable.ic_warning)
                .setCancelable(false)
                .setNegativeButton("Exit"
                ) { _, _ -> finish() }
        val alert = builder.create()
        alert.show()
    }
}